Prepare libQuotient for operating without being able to reach the server

TobiasFella opened the pull request.

TobiasFella committed:
Get rid of nested QEventLoop when loading pickling key

TobiasFella commented:
Great job, tobias, create a merge request that does everything *except* the thing it says it does :man_facepalming:

TobiasFella committed:
Add Connection::ready signal

This is important for clients to know the point from which it is safe to load state without
risking problems due to the e2ee machinery not being ready yet.

TobiasFella committed:
Connection::assumeIdentity: Set UserId and DeviceId and completeSetup before connecting to the server

This allows libQuotient-based clients to operate with the cached state (and in the future, room events)
when there is no network connection. It's losing the check whether userid and deviceid correspond to the given token,
but I don't see a strong need for that.

TobiasFella commented:
Slightly hijacked the merge request to do the rest of what I want to do; see individual commits for details

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -18,102 +18,109 @@
 
-Expected<PicklingKey, QKeychain::Error> setupPicklingKey(const QString& id,
-                                                         bool mock)
+void setupPicklingKey(const QString& id, bool mock, std::function<void(Expected<PicklingKey, QKeychain::Error>)> then)
```
Can you return a `QFuture<Expected<>>` here instead of adding a function parameter? Or if you want I can do it in-branch.

TobiasFella replied:
I tried doing that; I have no idea what I'm doing though and it just hangs in some future code. I'd appreciate if you could implement it

KitsuneRal commented on Quotient/connectionencryptiondata_p.cpp:
```diff
@@ -18,102 +18,109 @@
+    readJob->start();
+}
+
+void ConnectionEncryptionData::setup(Connection* connection, std::function<void(std::optional<std::unique_ptr<ConnectionEncryptionData>>)> then, bool mock)
```
...and then this function can be something like `return setupPicklingKey(...).then([] (Expected<...>) /*-> std::optional<...> */ { ... });` - or you can even reuse the `QFuture`'s cancelled state and strip `std::optional`, instead cancelling the future when the setup didn't work out. Again, I understand it's a change of paradigm so I can do it if you are not sure.

KitsuneRal reviewed (changes requested):
Thanks for reaching this corner - generally it looks good but I'd prefer futures to callbacks these days. It might be just me, so I'm open to discuss the benefits of either approach.

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -320,13 +320,15 @@ void Connection::Private::completeSetup(const QString& mxId, bool mock)
+                useEncryption = false;
+                emit q->encryptionChanged(false);
+            }
+            emit q->ready();
```
Right now, `TestOlmAccount` fails on my machine because `ready()` comes later than `connected()`, and the test uses `connected()` to proceed with its E2EE stuff, which is not there yet. I like the new signal name but I think having both signals together would confuse client authors. Should we keep `Connection::ready()` as the only recommended signal for clients, and deprecate `connected()` but fire it next to `ready()` in the meantime?

TobiasFella replied:
yes, makes sense

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -320,13 +320,15 @@ void Connection::Private::completeSetup(const QString& mxId, bool mock)
+        }, mock);
     } else
         qCInfo(E2EE) << "End-to-end encryption (E2EE) support is off for"
                      << q->objectName();
```
Also, shouldn't `ready()` be emitted here as well?

KitsuneRal committed:
Use futures instead of std::function() callbacks

KitsuneRal committed:
CI: Bump Qt Keychain to 0.14.3

Linux pipelines fail mysteriously since the rewrite, with the keychain
job getting deleted before yielding the result - maybe the newest
Qt Keychain would be better?

KitsuneRal committed:
CI: Now use the correct tag for Qt Keychain

KitsuneRal committed:
Replace code comments with logging

KitsuneRal committed:
Fix failing tests on Linux with Qt 6.4

Qt 6.4 doesn't handle QFutures with move-only results well.
The workaround for now is to pass the reference to the ultimate
move-only result (i.e. Connection::Private::encryptionData) into
all the continuations as a captured variable. Once we require
at least Qt 6.5 this workaround can be removed and normal QFuture
logic used to pass defered results around.

KitsuneRal commented:
The last round and then merge.
